File: Checkout/C++/get_order.cpp
/**
  Uses Openssl and cpprestsdk (tried on Arch linux)
  Run with: g++ -std=c++17 -lcpprest -lcrypto -lssl get_order.cpp -o get_order && ./get_order
*/

#include <cpprest/http_client.h>
#include <cpprest/json.h>
#include <iostream>
#include <fstream>
#include <string>

#include "SveaAuth.h"

int main() {
    std::cout << "Running GET request for Checkout (C++)" << std::endl;
    // Credentials
    std::string merchant_id = "CHECKOUT_MERCHANT_ID";
    std::string secret_word = "CHECKOUT_SECRET_KEY";
    std::string order_id = "";

    std::ifstream orderFile("./created_order_id.txt");
    if (orderFile.is_open()) {
        std::getline(orderFile, order_id);
        orderFile.close();
        //std::cout << "Using OrderId: " << order_id << std::endl;
    } else {
        std::cerr << "Failed to open ../created_order_id.txt. Ensure the file exists and is readable." << std::endl;
        return 1;
    }

    SveaAuth sveaAuthInstance;
    sveaAuthInstance.setMerchantId(merchant_id);
    sveaAuthInstance.setSecretWord(secret_word);

    web::http::http_request my_headers = sveaAuthInstance.get_request_headers();

    web::http::client::http_client_config config;

    // Could also use PA api if order is finalized
    web::http::client::http_client client("https://checkoutapistage.svea.com/api/orders/" + order_id, config);
    //web::http::client::http_client client("https://paymentadminapistage.svea.com/api/v1/orders/" + order_id, config);

    web::http::http_request request(web::http::methods::GET);
    for (const auto& header : my_headers.headers()) {
        request.headers().add(header.first, header.second);
    }

    web::http::http_response response = client.request(request).get();
    //std::cout << "Response status: " << response.status_code() << std::endl;

    if (response.status_code() == web::http::status_codes::OK) {
        web::json::value json_resp = response.extract_json().get();
        //std::cout << json_resp.serialize() << std::endl;
        std::cout << "Success!" << std::endl;
    } else {
        std::cout << "Failed..." << std::endl;
    }

    return 0;
}

File: Checkout/Php/get_order.php
<?php

try {
    $orderId = trim(file_get_contents('./created_order_id.txt'));
    if (!$orderId) {
        throw new Exception("OrderId is empty in the file.");
    }
    //echo "Using OrderId: $orderId\n";
} catch (Exception $e) {
    echo "Failed to read OrderId from file: " . $e->getMessage() . "\n";
    exit;
}

$merchantId = "CHECKOUT_MERCHANT_ID";
$secretWord = "CHECKOUT_SECRET_KEY";

class TestClass {

    public function get_request_headers($requestBody = "", $extraHeaders = null) {
        date_default_timezone_set('UTC');
        $timestamp = date("Y-m-d H:i:s");
        $token = $this->get_auth_token($timestamp, $requestBody);
        $headers = array("Authorization: Svea $token", "Timestamp: $timestamp");
        if ($extraHeaders) {
            $headers = array_merge($headers, $extraHeaders);
        }
        return $headers;
    }

    private function get_auth_token($timestamp, $requestBody = "") {
        $payloadJson = json_encode($requestBody); // Convert payload to JSON
        $payloadJson = "";
        $digest = $payloadJson . $GLOBALS['secretWord'] . $timestamp;
        $h = hash("sha512", $digest);
        $auth = $GLOBALS['merchantId'] . ":" . $h;
        $token = base64_encode($auth);
        return $token;
    }

    public function createTimestamp()
    {
        return gmdate('Y-m-d H:i');
    }
    
    public function createAuthorizationToken()
    {
        $timestamp = $this->createTimestamp();
        $authToken = base64_encode($GLOBALS['merchantId'] . ':' .
            hash('sha512', "" . $GLOBALS['secretWord'] . $timestamp));
        return $authToken;
    }
}

echo "Running GET request for Checkout (Php)\n";
$testInstance = new TestClass();
$myHeaders = $testInstance->get_request_headers(array('Content-Type' => 'application/json'));

// This also works
//$myHeaders = array();
//$myHeaders[] = 'Content-type: application/json';
//$myHeaders[] = 'Authorization: Svea ' . $testInstance->createAuthorizationToken();
//$myHeaders[] = 'Timestamp: ' . $testInstance->createTimestamp();
//$myHeaders[] = 'Expect:';

// $url = "https://paymentadminapistage.svea.com/api/v1/orders/" . $orderId;
$url = "https://checkoutapistage.svea.com/api/orders/" . $orderId;
$payload = array();

//echo "Fetching order: " . $orderId . "\n";
//echo "headers: " . $myHeaders[0] . "\n";

$ch = curl_init($url);
curl_setopt($ch, CURLOPT_HTTPHEADER, $myHeaders);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($ch, CURLOPT_HEADER, 1);
curl_setopt($ch,CURLOPT_SSL_VERIFYPEER, false);
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'GET');
//curl_setopt($ch, CURLOPT_VERBOSE, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, "");
$response = curl_exec($ch);
curl_close($ch);

//echo "Response: " . $response . "\n";

$httpcode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

if ($httpcode == 200) {
    echo "Success!\n";
} else {
    echo "Failed...\n";
}

?>
